### ISSUE TITLE
Convert importlib.abc to use typing.Protocol

### ISSUE THREAD

**Issue by brettcannon** [python/cpython#82963]
BPO | [38782](https://bugs.python.org/issue38782)
--- | :---
Nosy | @warsaw, @brettcannon, @ilevkivskyi, @ZackerySpytz, @nanjekyejoannah
PRs | <li>python/cpython#21245</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-11-12.22:47:40.242>
labels = ['type-feature', 'library', '3.9']
title = 'Convert importlib.abc to use typing.Protocol'
updated_at = <Date 2021-10-20.21:07:21.258>
user = 'https://github.com/brettcannon'
```

bugs.python.org fields:
```python
activity = <Date 2021-10-20.21:07:21.258>
actor = 'barry'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2019-11-12.22:47:40.242>
creator = 'brett.cannon'
dependencies = []
files = []
hgrepos = []
issue_num = 38782
keywords = ['patch']
message_count = 1.0
messages = ['356501']
nosy_count = 5.0
nosy_names = ['barry', 'brett.cannon', 'levkivskyi', 'ZackerySpytz', 'nanjekyejoannah']
pr_nums = ['21245']
priority = 'low'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue38782'
versions = ['Python 3.9']
```

</p></details>

**Comment by JelleZijlstra** [comment]
Protocols are a better conceptual fit for these classes, but wouldn't this change break backwards compatibility? For example, protocols don't support `.register()`.

**Comment by brettcannon** [comment]
Probably, which is why I figured the protocols would be separate since the ABCs also have default implementations that the protocols wouldn't have (I just forgot to update the issue with my latest thinking 😅). I had not decided if they should be in a separate module or just stick them in `importlib.abc`.

**Comment by brettcannon** [comment]
> protocols don't support `.register()`

Doesn't using `ABCMeta` get `Protocol` subclasses access to `register()`?

https://github.com/python/cpython/blob/f00512db20561370faad437853f6ecee0eec4856/Lib/typing.py#L2043

https://github.com/python/cpython/blob/f00512db20561370faad437853f6ecee0eec4856/Lib/typing.py#L1926

**Comment by ilevkivskyi** [comment]
Yes, IIRC protocols a supposed to be "drop in" replacement for ABCs, in particular `.register()` support is tested here https://github.com/python/cpython/blob/f00512db20561370faad437853f6ecee0eec4856/Lib/test/test_typing.py#L3765 That said however, most type-checkers don't support `.register()` well (or at all).

**Comment by brettcannon** [comment]
> That said however, most type-checkers don't support `.register()` well (or at all).

It would just be for the legacy ABC support. My expectation is the current ABCs will inherit from the protocols, but that most code will type to the protocols themselves and the type checkers will rely on that. But I would still have to either leave the abstract methods that are already there on the ABCs or use `@abc.abstractmethod` on the protocol classes themselves. Otherwise the protocols might as well be kept separate, in which case it's a question of where to put them in the importlib namespace.

Since I have not written it down, the protocols/types will be for:

- `sys.meta_path` finders
- `sys.path` finders
- Loaders
- `sys.path_hook` callables
